modules/asr/Cyberon/DSpotter_asr.c

```diff
@@ -0,0 +1,196 @@
+    DBG_TRACE("DSpotter version %s\r\n", DSpotterHL_GetVer());
+
+#ifdef UART_DUMP_RECORD
+    DBG_TRACE("It will send record data to UART. The format is four bytes audio data and one byte checksum.\r\n");
+    DBG_TRACE("\r\n");
+#endif
+
+    devmem_ctx = devmem;
+
+    oDSpotterInitData.nInitDataVer = DSPOTTER_INIT_DATA_VER;
+    oDSpotterInitData.nInitDataSize = (uint8_t)sizeof(DSpotterInitData);
+    oDSpotterInitData.nMaxCommandTime = MAX_COMMAND_TIME;
+    oDSpotterInitData.nCommandStageTimeout = COMMAND_STAGE_TIMEOUT;
+    oDSpotterInitData.nCommandStageFlowControl = COMMAND_STAGE_MULTI_COMMAND;
+    oDSpotterInitData.byAGC_Gain = (uint8_t)(VOLUME_SCALE_RECONG / 100);
+    oDSpotterInitData.bOneShotMode = true;
+    oDSpotterInitData.bExternalFlashModel = true;
+
+    nMemSize = DSpotterHL_GetMemoryUsage((const uint32_t *)model, &oDSpotterInitData);
+    DBG_TRACE("The DSpotter memory usage is %d.\r\n", nMemSize);
+    g_lpbyDSpotterMem = devmem_malloc(devmem_ctx, nMemSize);
+    if (g_lpbyDSpotterMem == NULL)
+    {
+        DBG_TRACE("devmem_malloc() fail!\r\n");
+        return NULL;
+    }
+
+    DBG_TRACE("DSpotterHL_Init\r\n");
+    nRet = DSpotterHL_Init((const uint32_t *)model, &oDSpotterInitData, g_lpbyDSpotterMem, nMemSize);
+    if (nRet != DSPOTTER_SUCCESS)
+    {
+        DBG_TRACE("DSpotterHL_Init() fail, error = %d!\r\n", nRet);
+        devmem_free(devmem_ctx, g_lpbyDSpotterMem);
+        g_lpbyDSpotterMem = NULL;
+        return NULL;
+    }
+
+    DBG_TRACE("The list of trigger word: \r\n");
+    nCount = DSpotterHL_GetDisplayCommandCount(DSPOTTER_HL_TRIGGER_STAGE);
+    for (int i = 0; i < nCount; i++)
+    {
+        DSpotterHL_GetDisplayCommand(DSPOTTER_HL_TRIGGER_STAGE, i, szCommand, sizeof(szCommand), &nCmdID);
+        DBG_TRACE("    %s, ID = %d\r\n", szCommand, nCmdID);
+    }
+
+    nCount = DSpotterHL_GetDisplayCommandCount(DSPOTTER_HL_COMMAND_STAGE);
+    if (nCount > 0)
+    {
+        DBG_TRACE("The list of command word: \r\n");
+        for (int i = 0; i < nCount; i++)
+        {
+            DSpotterHL_GetDisplayCommand(DSPOTTER_HL_COMMAND_STAGE, i, szCommand, sizeof(szCommand), &nCmdID);
+            DBG_TRACE("    %s, ID = %d\r\n", szCommand, nCmdID);
+        }
+    }
+    DBG_TRACE("\r\n");
+
+    return (asr_port_t)100;
+}
+
+asr_error_t asr_get_attributes(asr_port_t *ctx, asr_attributes_t *attributes)
+{
+    return ASR_NOT_SUPPORTED;
+}
+
+asr_error_t asr_process(asr_port_t *ctx, int16_t *audio_buf, size_t buf_len)
+{
+#ifdef UART_DUMP_RECORD
+    uint8_t byaTxBuffer[DSPOTTER_FRAME_SAMPLE*sizeof(int16_t)*3/2];
+    int nTransferSize;
+
+    if (buf_len < DSPOTTER_FRAME_SAMPLE)
+    {
+        nTransferSize = Convert2TransferBuffer((const uint8_t*)audio_buf, buf_len*sizeof(int16_t), byaTxBuffer, sizeof(byaTxBuffer), eFourByteDataOneChecksum);
+        rtos_uart_tx_write(uart_tx_ctx, byaTxBuffer, (uint32_t)nTransferSize);
+    }
+#else
+    if (++g_nRecordFrameCount % 100 == 0)
```

Comment:
It's a bit dodgy not introducing a new scope for the macro below.  What happens if the macro includes more than one C statement?

Reply:
Added a new scope.